Step in the direction of RA2 paratroopers

matija-hustic opened the pull request:
`Parachute` effect replaced by an in-world actor implementing the `Parachute` activity. That activity grants a `parachute` upgrade, changes the actor's targetability and moves the actor towards the ground. Once on the ground, the `parachute` upgrade is revoked.

The `parachute` upgrade is consumed by `WithParachute`, an upgradable trait that renders the parachute (similar to `WithIdleOverlay`) and the shadow (using `IRender`).

Parachuted actors are made uncrushable and unable to crush.

Two further small adjustments have been made to the `SetPosition` placement functions on `Mobile` and `Crate` traits.

Edit: Fixes #2988.

pchote commented on OpenRA.Mods.Common/Activities/Parachute.cs:
```diff
@@ -0,0 +1,77 @@
+
+		public Parachute(Actor self, WPos dropPosition)
+		{
+			var um = self.TraitOrDefault<UpgradeManager>();
```
In general, activities may be queued and not run until an arbitrary time after they are constructed.  It is therefore best practice to not do anything that changes world state from their constructor.  You can call these from the first `Tick()`.

matija-hustic replied:
Fixed locally. I'll push after other remarks.

LipkeGu commented:
we need this for TS to o.o" -> drop pods

abcdefg30 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -573,6 +576,9 @@ public bool CanMoveFreelyInto(CPos cell, Actor ignoreActor = null, bool checkTra
 		{
+			// Only make actor crush if it is on the ground
+			if (self.CenterPosition.Z != 0)
+				return;
```
Can we have a new line after this?

abcdefg30 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -581,6 +587,9 @@ public void EnteringCell(Actor self)
 		{
+			// Only make actor crush if it is on the ground
+			if (self.CenterPosition.Z != 0)
+				return;
```
Same here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -125,8 +125,39 @@ public void Tick(Actor self)
+			}
+		}
+
+		private void SetLocation(Actor self, CPos cell, SubCell subCell = SubCell.Any)
```
Seems like you forgot to comment here as you did for the other methods?

matija-hustic replied:
I didn't forget, I just thought clarification wasn't necessary with this function. I'll add the comments tho.

Phrohdoh replied:
No explicit private.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -155,21 +186,11 @@ public bool CanEnterCell(CPos a, Actor ignoreActor = null, bool checkTransientAc
 		{
+			// Crate can only be crushed if it is not in the air or underground
+			if (CenterPosition.Z != 0)
+				return false;
```
And maybe a new line here too?

abcdefg30 commented on OpenRA.Mods.Common/Activities/Parachute.cs:
```diff
@@ -0,0 +1,77 @@
+
+			var p = self.Info.Traits.GetOrDefault<ParachutableInfo>();
+			if (p != null)
+				fallVector = new WVec(0, 0, p.FallRate);
```
Can we get a default for this or require the `Parachutable` trait?
If a unit doesn't have the `Parachutable` trait it'll just stay in air all the time (in this PR).

penev92 replied:
This is in `Activities/Parachute`, so it should require the `Parachutable` trait.
Otherwise we can have a `Freefall` activity for actors that don't have it :D

matija-hustic replied:
So do I seriously go for this `Freefall`? :D

abcdefg30 commented:
You should also set the paradrop-traits at `^Tank` in ra up or remove them from `^Tank` as we don't have the `ArmorDrop` support power anymore.

abcdefg30 commented:
We should also fix infantry going prone in air and the death animation somewhen.

matija-hustic commented:
Fixed as commented and rebased. Only remains to see about
* the `Parachutable` trait being a prerequisite / adding the `Freefall` activity and
* the paratrooper animation (not going prone + death).

Maybe I can some more input on what would be the right way to go with this PR and what should be left for an additional PR?

GraionDilach commented:
I did a quick glampse on the code and while I do appreciate this there's one thing I noticed:

RA2 parachute has a closing sequence as well. It doesn't disappear like the RA1 one, but fall to the ground then play a decayanim. Notsure how to implement this via the upgrade tho.

GraionDilach commented on OpenRA.Mods.Common/Traits/Render/WithParachute.cs:
```diff
@@ -0,0 +1,116 @@
+		[Desc("Prachute opening sequence.")]
+		public readonly string OpeningSequence = "open";
+
+		[Desc("Prachute idle sequence.")]
```
Typo (there's also one at ShadowSequence's description along with these two).

matija-hustic commented:
@GraionDilach, it isn't too problematic to implement the closing sequence. The only question is - seeing how the closing sequence needs to play while the unit is already able to move around, should the animation move with him or play out on the drop location?

GraionDilach commented:
In RA2 it moved along with it, it was attached to the unit like the EMP or the mindcontrol animation.

(Actually, WW even had a reason to do that although they made the anim entirely to work exactly with their flying height. The unit had a doesitstillexist check upon the parachute anim and the result was used to decide between applying [General]ParachuteMaxFallRate or NoParachuteMaxFallRate, but since FallRate is directly customizable here, your solution is better than theirs.)

penev92 commented:
I think moving to the north would look bad if the parachute moves with the infantry (and keeps its relative position), so probably drop the parachute when you hit the ground (a.k.a. play it where they landed). But that would be trickier at best, on top of having people go "that's not what the original did!", so...

matija-hustic commented:
So which one should it be? With the current code it would follow the unit.

GraionDilach commented:
I'd say current will do in this case.

obrakmann commented on OpenRA.Mods.Common/Activities/Parachute.cs:
```diff
@@ -0,0 +1,96 @@
+
+			var um = self.TraitOrDefault<UpgradeManager>();
+			if (um != null)
+				um.GrantUpgrade(self, "parachute", this);
```
It would be nice to un-hardcode the name of the upgrade and make it into a field in `ParachutableInfo`.

Also, you need the `UpgradeManager` twice, so it's probably already worth caching it in the constructor.

obrakmann commented on OpenRA.Mods.Common/Activities/Parachute.cs:
```diff
@@ -0,0 +1,96 @@
+			if (um != null)
+				um.GrantUpgrade(self, "parachute", this);
+
+			var tpu = self.TraitOrDefault<TargetableParachutableUnit>();
```
Please cache this, too

obrakmann commented on OpenRA.Mods.Common/Activities/Parachute.cs:
```diff
@@ -0,0 +1,96 @@
+			fallVector = new WVec(0, 0, p.FallRate);
+
+			// Place the actor and retrieve its visual position (CenterPosition)
+			self.Trait<IPositionable>().SetPosition(self, dropPosition);
```
and `IPositionable` as well

obrakmann commented on OpenRA.Mods.Common/Traits/ParaDrop.cs:
```diff
@@ -82,7 +82,11 @@ public void Tick(Actor self)
+			self.World.AddFrameEndTask(w =>
+			{
+				w.Add(a);
+				a.QueueActivity(new Activities.Parachute(a, self.CenterPosition));
```
You can drop the `Activities` here, you're already including the namespace.

obrakmann commented on OpenRA.Mods.Common/Traits/EjectOnDeath.cs:
```diff
@@ -66,7 +66,11 @@ public void Killed(Actor self, AttackInfo e)
+					self.World.AddFrameEndTask(w =>
+					{
+						w.Add(pilot);
+						pilot.QueueActivity(new Activities.Parachute(pilot, cp));
```
I think you can clean up the `using`s here, too, and then remove the `Activities`.

Bonus points if you add a `Desc()` that states the traits the `PilotActor` needs in the Info class.

obrakmann commented:
And I'm afraid these changes call for an upgrade rule.

matija-hustic commented:
I'll put the upgrade rule and rest of YAML changes at the very end as @Mailaender once suggested. Otherwise I'll probably need to rebase alot.

Now also prevents paratroopers from idling while in air.

obrakmann commented on OpenRA.Mods.Common/Activities/Parachute.cs:
```diff
@@ -0,0 +1,99 @@
+		WPos currentPosition;
+		WVec fallVector;
+		bool triggered = false;
+		UpgradeManager um;
```
You can make all of these cached traits `readonly`. (And iirc you'll have to move them to the top then to conform to the coding style. Run StyleCop to see if it complains about anything.)

obrakmann commented on OpenRA.Mods.Common/Traits/EjectOnDeath.cs:
```diff
@@ -8,6 +8,7 @@
 #endregion
 
+using OpenRA.Mods.Common.Activities;
 using OpenRA.Mods.Common.Effects;
```
Unless I missed something, I think you don't need the `Effects` anymore.

obrakmann commented on OpenRA.Mods.Common/Traits/Render/WithParachute.cs:
```diff
@@ -0,0 +1,133 @@
+	{
+		Animation shadow;
+		AnimationWithOffset anim;
+		WithParachuteInfo info;
```
These can all be `readonly`, too.

obrakmann commented on OpenRA.Mods.Common/Traits/Render/WithInfantryBody.cs:
```diff
@@ -113,16 +113,17 @@ public virtual void Tick(Actor self)
 				DefaultAnimation.PlayRepeating(NormalizeInfantrySequence(self, Info.MoveSequence));
 			}
+			else if (((state == AnimationState.Moving || dirty) && !move.IsMoving)
+			 || ((state == AnimationState.Idle || state == AnimationState.IdleAnimating) && !self.IsIdle))
```
Looks like a space instead of a tab at the beginning of the line.

obrakmann commented:
Aside from those codestyle niggles, this is looking otherwise fine to me. I'll give my :+1: when the upgrade rules are added.

matija-hustic committed:
Step in the direction of RA2 paratroopers.

matija-hustic commented:
* Replaced `TargetableParachutableUnit` with an upgradable `TargetableUnit`
* Added an upgrade rule
* Upgraded all mods
* Rebased and squashed

abcdefg30 commented on OpenRA.Mods.Common/Traits/TargetableUnit.cs:
```diff
@@ -52,5 +63,26 @@ public virtual IEnumerable<WPos> TargetablePositions(Actor self)
+				if (info.UpgradedTargetTypes.Contains("!"))
+					targetTypes = info.UpgradedTargetTypes.Where(utt => !utt.StartsWith("!")).ToArray();
+				else
+					targetTypes = targetTypes.Except(info.UpgradedTargetTypes.Where(utt => utt.StartsWith("!")).Select(utt => utt.Substring(1)))
```
Sorry if I missunderstood something here, but aren't you checking for `info.UpgradedTargetTypes.Contains("!")` above, so `info.UpgradedTargetTypes.Where(utt => utt.StartsWith("!")).Select(...` is redundant here?
(As something that doesn't contain a "!" shouldn't start with it :p)

matija-hustic replied:
`info.UpgradedTargetTypes` is a collection of strings, `utt` is a string from this collection

`info.UpgradedTargetTypes.Contains()` checks if a string is present in the collection
`utt.StartsWith()` checks if a string begins by a substring

:)

obrakmann replied:
No, that's correct. The first `if` checks whether a *single* ! exists in `UpgradedTargetTypes`. In that case, `TargetTypes` will be *replaced* with all entries in `UpgradedTargetTypes` that don't start with an !.

The `else` case instead *modifies* the existing `TargetTypes` and removes all entries that start with an !, and adds all others.

obrakmann commented:
Looks ok to me, and works as expected in-game. :+1:

pchote commented on OpenRA.Mods.Common/Traits/TargetableUnit.cs:
```diff
@@ -20,20 +21,30 @@ public class TargetableUnitInfo : ITraitInfo, ITargetableInfo
+
+		[Desc("List of targets to add or remove when at least one upgrade is granted. ",
+			"Prefix target types to remove with an exclamation point (!). ",
+			"Use the exclamation point alone to remove all target types.")]
```
We tend to avoid magic values where possible, so lets make this a full list of types instead, please.  You could default this to null, and then have it fall back to TargetTypes in that case.

reaperrr commented:
Needs a rebase.

matija-hustic commented:
Changed as commented. Rebased.

GraionDilach commented:
I don't like the TargetableUnit upgrade implementation. It's okay for this particular case, but it limits us in the long run. There will be cases when multiple TargetType upgrades each having separate effects will be desired. (RA2 paratroopers upgrade the TargetType to air-only while Chrono Legionnaires should upgrade it to Temporal-only, reimplementing the tiberiumheal and tiberiumproof veteraneffects in TS due to both effects rely on TargetTypes in OpenRA to apply their effects as well).

It should be refactored to allow using IdentifierSuffixes in a separate PR.

Mailaender commented on OpenRA.Mods.Common/Traits/Render/WithParachute.cs:
```diff
@@ -0,0 +1,134 @@
+		public readonly string OpeningSequence = "open";
+
+		[Desc("Parachute idle sequence.")]
+		public readonly string Sequence = "idle";
```
You forgot all the `[SequenceReference("Image")]` during the refactor. See https://github.com/openra/openra/pull/7709.

matija-hustic commented:
Fixed as @Mailaender commented. Rebased.

@GraionDilach, I'm not sure I followed what you meant with you remark on `TargetableUnit`, but if it can be done in a separate PR, as you said, I'd say go for it.

Mailaender commented:
This spits out a lot of MiniYaml errors and warnings: https://travis-ci.org/OpenRA/OpenRA/builds/68995836#L1177

matija-hustic commented:
Does d2k use parachuting units?

Mailaender commented:
Yes, but without parachute sequence in the Carryall drop.

matija-hustic commented:
Is that how things ought to be or do they need a parachute sequence?

Mailaender commented:
They need to have none. My previous traits did not hard-code the sequences and had null checks to omit the animations in that case.

pchote commented:
The carryall drop will be removed if we ever proceed to the next stage of the d2k mod.

matija-hustic committed:
Manually upgraded YAMLs.

matija-hustic commented:
OK, so basically the lint checks, that have been added since this PR was open, have forced me to do some new stuff. Seeing how `WithParachute` doesn't support rendering the shadow as just the actor's usual sprite with the shadow palette, any parachutable actors that have relied on their shadow being rendered this way will no longer have a shadow when automatically upgrading the mods. I've added the nicer shadow `parach-shadow` manually for crates and d2k paratroops (this doesn't happen automatically as it potentially requires copying the image file and adding a sequence entry).

And, parachuting without a parachute sequence is now also handled with null-checks as before.

Mailaender commented:
Needs a rebase.

Mailaender commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -1306,11 +1306,11 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 
 							if (ibfNode.Any())
 								rh.Value.Nodes.RemoveAll(n => ibfNode.Contains(n));
-								Console.WriteLine("The 'ImagesByFullness' property from the removed RenderHarvester trait has been");
```
You fixed the indention error the wrong way here. https://github.com/OpenRA/OpenRA/pull/8644/ did it properly already.

matija-hustic replied:
This was automatic when I rebased last night. It'll go away when I rebase now.

Mailaender replied:
Be sure to leave a message when you do. I saw some of your pull request being silently rebased so they weren't merged already, because no one noticed.

matija-hustic commented:
Rebased.

Mailaender commented on OpenRA.Mods.Common/Traits/TargetableUnit.cs:
```diff
@@ -9,28 +9,33 @@
 {
 	[Desc("Actor can be targeted.")]
-	public class TargetableUnitInfo : ITraitInfo, ITargetableInfo
+	public class TargetableUnitInfo : UpgradableTraitInfo, ITraitInfo, ITargetableInfo
```
This conflicts with https://github.com/OpenRA/OpenRA/pull/8662 and somehow broke things. The defending allied units in the shellmap don't retaliate anymore.

pchote commented:
What's the situation with this?  It seems a bit unfair that newer prs that conflict with this are being preferentially merged.

matija-hustic commented:
@GraionDilach and @Mailaender didn't how the `TargetableUnit` was modified so I'm waiting for #8662 to get merged before I rebase and use it.

GraionDilach commented:
#8662 is a step towards removing all `TargetableUnit` subclasses. That's something I feel preferable over the implementation provided here.

matija-hustic committed:
Automatically upgraded YAMLs.

matija-hustic commented:
Rebased and using #8662. Shellmap works normally.

GraionDilach commented on OpenRA.Mods.Common/Traits/Parachutable.cs:
```diff
@@ -30,25 +30,11 @@ class ParachutableInfo : ITraitInfo
-		public readonly string ShadowPalette = null;
+		[UpgradeGrantedReference]
+		[Desc("Upgrade to grant to this actor when parachuting. Normally used to render the parachute using the WithParachute trait.")]
+		public readonly string ParachuteUpgrade = "parachute";
```
Could this be a list, please?

GraionDilach commented on mods/ra/rules/defaults.yaml:
```diff
@@ -648,10 +658,17 @@
+		OpeningSequence: open
+		ShadowImage: parach-shadow
+		ShadowSequence: idle
+	UpgradeManager:
```
Something implies me this was caused by auto-upgrading - `^Soldier` inherits both two required traits from `^ExistsInWorld`.

matija-hustic replied:
This is a `^Crate` YAML here. Expand the code above ^^

GraionDilach replied:
Oh, sorry. I misread it as the D2K infantry for some reason.

GraionDilach commented:
I think I can safely give my :+1: when the above nits are resolved.

obrakmann commented:
Sorry, needs a rebase.

matija-hustic commented:
Rebased.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/db824811bcabdc6f2cf5c011790022ea678abf0f%5E...db824811bcabdc6f2cf5c011790022ea678abf0f)

penev92 commented on mods/d2k/rules/defaults.yaml:
```diff
@@ -225,6 +227,11 @@
+		UpgradeTypes: parachute
+		UpgradeMinEnabledLevel: 1
+		ShadowImage: parach-shadow
+		ShadowSequence: idle
```
This somehow broke D2k:
>Unit `parach-shadow` does not have a sequence named `idle`

GraionDilach replied:
The shadow is in TS SHP format instead of TD SHP format. >_>

matija-hustic replied:
Godamn it. What do you guys use for SHP editing? I've downloaded this one program and I can't get anything done with it properly

matija-hustic replied:
Altho, I don't remember messing around with this particular SHP. It was just a copy paste if I remember correctly

GraionDilach replied:
I tend to go XCC Mixer (being part of XCC Utilities) from http://xhp.xwis.net/utilities/ due to that being practically the swiss knife WW modding tool. It has some bugs - only read pals within mixes, TS and RA2 folders are read from registry - but people can easily create a mix and feed it's folder as TD or RA1 install folder instead. It works under Wine.
